src/runtime/security/security.thrift

```diff
@@ -0,0 +1,46 @@
+include "../../dsn.thrift"
+
+namespace cpp dsn.security
+
+// negotiation process:
+//
+//                       client                              server
+//                          | ---   SASL_LIST_MECHANISMS    --> |
+//                          | <-- SASL_LIST_MECHANISMS_RESP --- |
+//                          | --    SASL_SELECT_MECHANISMS  --> |
+//                          | <-- SASL_SELECT_MECHANISMS_OK --- |
+//                          |                                   |
+//                          | ---      SASL_INITIATE        --> |
+//                          |                                   |
+//                          | <--      SASL_CHALLENGE       --- |
+//                          | ---      SASL_RESPONSE        --> |
+//                          |                                   |
+//                          |              .....                |
+//                          |                                   |
+//                          | <--      SASL_CHALLENGE       --- |
+//                          | ---      SASL_RESPONSE        --> |
+//                          |                                   | (authentication will succeed
+//                          |                                   |  if all chanllenges passed)
+//                          | <--        SASL_SUCC          --- |
+// (client won't response   |                                   |
+// if servers says ok)      |                                   |
+//                          | ---        RPC_CALL          ---> |
+//                          | <--        RPC_RESP          ---- |
+
+enum negotiation_status {
+    INVALID,
+    SASL_LIST_MECHANISMS,
+    SASL_LIST_MECHANISMS_RESP,
+    SASL_SELECT_MECHANISMS,
+    SASL_SELECT_MECHANISMS_OK,
+    SASL_INITIATE,
+    SASL_CHALLENGE,
+    SASL_RESPONSE,
+    SASL_SUCC,
+    SASL_AUTH_FAIL
+}
+
+struct negotiation_message {
```

Comment:
What is massage used for?

Reply:
rpc message exchanged between server and client

Reply:
rpc message exchanged between client and server

Reply:
It's a request from client to server or server to client, or both?
How can I distinguish this message is a request or a response?

Reply:
both. A rpc session is client or server. Client will send request and receive response. It's easy to distinguish them.

Reply:
I see client send `negotiation_message` to server, how about call this structure `negotiation_request`? and add another structure named `negotiation_response`, `message` makes me confused that whether it is a request or response.

Reply:
It's hard to spilt `negotiation_message` into `negotiation_request` and `negotiation_response`. Because it's hard to distinguish which one is request(or response) from `SASL_INITIATE`、`SASL_CHALLENGE` and `SASL_RESPONSE`. They are all negotiation_message. And the logic of `client_negotiation` and `server_negotiation` will be too complex, because they should distinguish which `negotiation_message` is a `negotiation_request` and `negotiation_response`